packages/invoices/src/InboundShipment/DetailView/api.ts

```diff
@@ -204,3 +218,121 @@ export const getInboundShipmentDetailViewApi = (
+  });
+};
+
+export const useInboundShipmentSelector = <T = Invoice>(
+  select?: (data: Invoice) => T
+): UseQueryResult<T, unknown> => {
+  const { id } = useParams();
+  const { api } = useOmSupplyApi();
+  const queries = getInboundShipmentDetailViewApi(api);
+  return useQuery(
+    ['invoice', id],
+    () => {
+      return queries.onRead(id);
+    },
+    { select }
+  );
+};
+
+const useOptimisticInboundUpdate = () => {
+  const { api } = useOmSupplyApi();
+  const queries = getInboundShipmentDetailViewApi(api);
+  const queryClient = useQueryClient();
+  const { id } = useParams();
+  return useMutation(queries.onUpdate, {
+    onMutate: async (patch: Partial<InboundShipment>) => {
+      await queryClient.cancelQueries(['invoice', id]);
+
+      const previousInbound: Invoice = queryClient.getQueryData([
+        'invoice',
+        id,
+      ]);
+
+      queryClient.setQueryData(['invoice', id], {
+        ...previousInbound,
+        ...patch,
+      });
+
+      return { previousInbound, patch };
+    },
+    onSettled: () => queryClient.invalidateQueries(['invoice', id]),
+    onError: (_, __, context) => {
+      queryClient.setQueryData(['invoice', id], context.previousInbound);
+    },
+  });
+};
+
+export const useInboundLines = () => {
+  const { sortBy, onChangeSortBy } = useSortBy<OutboundShipmentSummaryItem>({
+    key: 'itemName',
+  });
+  const onSort = (column: Column<OutboundShipmentSummaryItem>) => {
+    onChangeSortBy({
+      key: column.key,
+      isDesc: sortBy.key === column.key ? !sortBy.isDesc : false,
+    });
+  };
+  const selectItems = useCallback(
+    (invoice: Invoice) => {
+      return inboundLinesToSummaryItems(invoice.lines).sort(
+        getDataSorter(
+          sortBy.key as keyof OutboundShipmentSummaryItem,
+          !!sortBy.isDesc
+        )
+      );
+    },
+    [sortBy]
+  );
+
+  const { data } = useInboundShipmentSelector(selectItems);
+
+  return { data, sortBy, onSort };
+};
+
+export const useDraftInbound = () => {
+  const queryClient = useQueryClient();
+  const { id } = useParams();
+  const { api } = useOmSupplyApi();
+  const queries = getInboundShipmentDetailViewApi(api);
+
+  const { data } = useInboundShipment();
+
+  const draft = data ? { ...data, items: [] } : placeholderInbound;
+  const { mutateAsync: optimisticUpdate } = useOptimisticInboundUpdate();
+  const { isLoading: isAddingItem, mutateAsync } = useMutation(
+    queries.onUpdate,
+    {
+      onSettled: () => queryClient.invalidateQueries(['invoice', id]),
+    }
+  );
+
+  const updateInvoice = async (patch: Partial<InboundShipment>) => {
+    return optimisticUpdate({ ...data, ...patch, items: [] });
+  };
+
+  const upsertItem = async (item: OutboundShipmentSummaryItem) => {
+    const itemIdx = draft.items.findIndex(i => i.id === item.id);
+    if (itemIdx >= 0) draft.items[itemIdx] = item;
+    else draft.items.push(item);
+
+    const result = await mutateAsync(draft);
+
+    return result;
+  };
```

Comment:
I think what would be best would be to separate out the mutations into individual hooks and use them closer to where they are used.

For example, in the `SidePanel` component, using `useInboundSelector` to select the specific parts of the invoice that are being displayed, and `useUpdateInbound` to get the mutate function to update rather than passing down props to do all this.

But wanted to just wait and get some feedback first!

Reply:
yes! agree on both counts - would be great to separate out the mutations, and yeah, good to wait for feedback